src/coreclr/jit/promotionliveness.cpp

```diff
@@ -0,0 +1,869 @@
+
+            if ((ssize_t)index < 0)
+            {
+                index = ~index;
+                if ((index > 0) && reps[index - 1].Overlaps(offs, size))
+                {
+                    index--;
+                }
+            }
+
+            while ((index < reps.size()) && (reps[index].Offset < offs + size))
+            {
+                Replacement& rep = reps[index];
+                bool         isFullFieldDef =
+                    isDef && (offs <= rep.Offset) && (offs + size >= rep.Offset + genTypeSize(rep.AccessType));
+                MarkIndex(baseIndex + 1 + (unsigned)index, isUse, isFullFieldDef, useSet, defSet);
+                index++;
+            }
+
+            bool isFullDefOfRemainder = isDef && (agg->UnpromotedMin >= offs) && (agg->UnpromotedMax <= (offs + size));
+            // TODO-CQ: We could also try to figure out if a use actually touches the remainder, e.g. in some cases
+            // a struct use may consist only of promoted fields and does not actually use the remainder.
+            MarkIndex(baseIndex, isUse, isFullDefOfRemainder, useSet, defSet);
+        }
+    }
+    else
+    {
+        unsigned offs  = lcl->GetLclOffs();
+        size_t   index = Promotion::BinarySearch<Replacement, &Replacement::Offset>(reps, offs);
+        if ((ssize_t)index < 0)
+        {
+            unsigned size             = genTypeSize(accessType);
+            bool isFullDefOfRemainder = isDef && (agg->UnpromotedMin >= offs) && (agg->UnpromotedMax <= (offs + size));
+            MarkIndex(baseIndex, isUse, isFullDefOfRemainder, useSet, defSet);
+        }
+        else
+        {
+            // Accessing element.
+            MarkIndex(baseIndex + 1 + (unsigned)index, isUse, isDef, useSet, defSet);
+        }
+    }
+}
+
+//------------------------------------------------------------------------
+// MarkIndex:
+//   Mark specific bits in use/def bit vectors depending on whether this is a use def.
+//
+// Parameters:
+//   index  - The index of the bit to set.
+//   isUse  - Whether this is a use
+//   isDef  - Whether this is a def
+//   useSet - The set of uses
+//   defSet - The set of defs
+//
+void PromotionLiveness::MarkIndex(unsigned index, bool isUse, bool isDef, BitVec& useSet, BitVec& defSet)
+{
+    if (isUse && !BitVecOps::IsMember(m_bvTraits, defSet, index))
+    {
+        BitVecOps::AddElemD(m_bvTraits, useSet, index);
+    }
+
+    if (isDef)
+    {
+        BitVecOps::AddElemD(m_bvTraits, defSet, index);
+    }
+}
+
+//------------------------------------------------------------------------
+// InterBlockLiveness:
+//   Compute the fixpoint.
+//
+void PromotionLiveness::InterBlockLiveness()
+{
+    bool changed;
+    do
+    {
+        changed = false;
+
+        for (BasicBlock* block = m_compiler->fgLastBB; block != nullptr; block = block->bbPrev)
+        {
+            m_hasPossibleBackEdge |= block->bbNext && (block->bbNext->bbNum <= block->bbNum);
```

Comment:
We know for sure we have proper bbNum ordering?

In cases with loops this will converge faster if you walk the blocks in post-order here, to maximize the propagation per iteration.

Reply:
Looks like the same goes for regular liveness so perhaps we can fix that too -- maybe do both as a follow-up?

Reply:
> We know for sure we have proper bbNum ordering?

Even if we don't it will just result in `m_hasPossibleBackEdge` being set to true which might result in one more iteration of the fixpoint computation than necessary.

When I did early liveness I considered computing strongly connected components and doing the localized fixpoint computation within each SCC, which should be the optimal thing. However, when I measured the actual throughput cost of the fixpoint computation I felt it wasn't worth more complexity, from https://github.com/dotnet/runtime/pull/79346#issuecomment-1345404345:

```scala
Base: 232865278205, Diff: 236405285056, +1.5202%

?PerBlockAnalysis@LiveVarAnalysis@@AEAA_NPEAUBasicBlock@@_N1@Z                                                                                                                                                                     : 616293246 : +30.93%  : 11.44% : +0.2647%
?fgMarkUseDef@Compiler@@AEAAXPEAUGenTreeLclVarCommon@@@Z                                                                                                                                                                           : 426414825 : +39.60%  : 7.91%  : +0.1831%
?fgInterBlockLocalVarLiveness@Compiler@@QEAAXXZ                                                                                                                                                                                    : 353468679 : +43.94%  : 6.56%  : +0.1518%
?fgPerBlockLocalVarLiveness@Compiler@@QEAAXXZ                                                                                                                                                                                      : 258048984 : +26.39%  : 4.79%  : +0.1108%
?lvaSortByRefCount@Compiler@@QEAAXXZ                                                                                                                                                                                               : 222829814 : +39.99%  : 4.14%  : +0.0957%
?GetSucc@BasicBlock@@QEAAPEAU1@IPEAVCompiler@@@Z                                                                                                                                                                                   : 142747231 : +11.06%  : 2.65%  : +0.0613%
?PostOrderVisit@ForwardSubVisitor@@QEAA?AW4fgWalkResult@Compiler@@PEAPEAUGenTree@@PEAU4@@Z                                                                                                                                         : 136636902 : NA       : 2.54%  : +0.0587%
?WalkTree@?$GenTreeVisitor@VLocalSequencer@@@@QEAA?AW4fgWalkResult@Compiler@@PEAPEAUGenTree@@PEAU4@@Z                                                                                                                              : 134894223 : NA       : 2.50%  : +0.0579%
??0AllSuccessorIterPosition@@QEAA@PEAVCompiler@@PEAUBasicBlock@@@Z                                                                                                                                                                 : 129181805 : +13.59%  : 2.40%  : +0.0555%
?fgComputeLifeTrackedLocalUse@Compiler@@QEAAXAEAPEA_KAEAVLclVarDsc@@PEAUGenTreeLclVarCommon@@@Z                                                                                                                                    : 127997140 : +37.14%  : 2.38%  : +0.0550%
?fgForwardSubStatement@Compiler@@AEAA_NPEAUStatement@@@Z                                                                                                                                                                           : 119968922 : +70.17%  : 2.23%  : +0.0515%
?fgMarkAddressExposedLocals@Compiler@@AEAA?AW4PhaseStatus@@XZ                                                                                                                                                                      : 119662088 : +72.72%  : 2.22%  : +0.0514%
?FindNextRegSuccTry@EHSuccessorIterPosition@@AEAAXPEAVCompiler@@PEAUBasicBlock@@@Z                                                                                                                                                 : 116051428 : +13.69%  : 2.15%  : +0.0498%
?WalkTree@?$GenTreeVisitor@V?$GenericTreeWalker@$00$0A@$0A@$00@@@@QEAA?AW4fgWalkResult@Compiler@@PEAPEAUGenTree@@PEAU4@@Z                                                                                                          : 111845378 : +56.69%  : 2.08%  : +0.0480%
?fgComputeLifeTrackedLocalDef@Compiler@@QEAA_NAEAPEA_KAEBQEA_KAEAVLclVarDsc@@PEAUGenTreeLclVarCommon@@@Z                                                                                                                           : 108064056 : +34.46%  : 2.01%  : +0.0464%
memset                                                                                                                                                                                                                             : 104530112 : +5.76%   : 1.94%  : +0.0449%
```
So around 0.26% in TP cost for the fixpoint computation from adding early liveness.

It would still be an interesting experiment, and RPO traversal is also simpler than computing full SCCs.